com.unity.shadergraph/Editor/Drawing/PreviewManager.cs

```diff
@@ -243,276 +282,443 @@ public bool HandleGraphChanges()
-            m_Messenger.ClearNodesFromProvider(this, m_NodesToUpdate);
-            var wasAsyncAllowed = ShaderUtil.allowAsyncCompilation;
-            ShaderUtil.allowAsyncCompilation = true;
+                // add each node to compile list if it needs a preview, is not already compiling, and we have room
+                // (we don't want to double kick compiles, so wait for the first one to get back before kicking another)
+                if (m_NodesCompiling.Count + nodesToCompile.Count < m_MaxNodesCompiling)
+                {
+                    foreach (var node in m_NodesNeedsRecompile)
+                    {
+                        if (node.hasPreview && node.previewExpanded && !m_NodesCompiling.Contains(node))
+                        {
+                            var renderData = GetPreviewRenderData(node);
+                            if (renderData == null) // non-active output nodes can have NULL render data (no preview)
+                                continue;
 
-            foreach (var node in m_NodesToUpdate)
-            {
-                if (node is IMasterNode && node == masterRenderData.shaderData.node && !(node is VfxMasterNode))
+                            nodesToCompile.Add(node);
+                            if (m_NodesCompiling.Count + nodesToCompile.Count >= m_MaxNodesCompiling)
+                                break;
+                        }
+                    }
+                }
+
+                // remove the selected nodes from the recompile list
+                m_NodesNeedsRecompile.ExceptWith(nodesToCompile);
+
+                // Reset error states for the UI, the shader, and all render data for nodes we're recompiling
+                m_Messenger.ClearNodesFromProvider(this, nodesToCompile);
+
+                // Force async compile on
+                var wasAsyncAllowed = ShaderUtil.allowAsyncCompilation;
+                ShaderUtil.allowAsyncCompilation = true;
+
+                // kick async compiles for all nodes in m_NodeToCompile
+                foreach (var node in nodesToCompile)
                 {
-                    UpdateMasterNodeShader();
-                    continue;
+                    if (node is IMasterNode && node == masterRenderData.shaderData.node && !(node is VfxMasterNode))
+                    {
+                        UpdateMasterNodeShader();
+                        continue;
+                    }
+
+                    Assert.IsFalse(!node.hasPreview && !(node is SubGraphOutputNode || node is VfxMasterNode));
+
+                    var renderData = GetPreviewRenderData(node);
+
+                    // Get shader code and compile
+                    var generator = new Generator(node.owner, node, GenerationMode.Preview, $"hidden/preview/{node.GetVariableNameForNode()}");
+                    BeginCompile(renderData, generator.generatedShader);
+
+                    // Calculate the PreviewMode from upstream nodes
```

Comment:
This is unrelated to your perf changes but why is this code here? Can't we just check it in HandleGraphChanges when nodes or edges actually change?

Reply:
I had exactly that comment in here before I cleaned up the code!
No idea why it is here....  could add a TODO: move this somewhere that makes more sense...

Reply:
Do or do not, there is no TODO. ;) We never address them and 3 years from now after 6 refactors somebody will have to spend time thinking about whether that TODO still makes any sense. It works, so /shrug.